core/src/main/java/com/linecorp/armeria/server/auth/HttpAuthService.java

```diff
@@ -66,52 +66,44 @@
         return newDecorator(ImmutableList.copyOf(requireNonNull(authorizers, "authorizers")));
     }
 
-    /**
-     * Creates a new instance that provides HTTP authorization functionality to {@code delegate}.
-     */
-    protected HttpAuthService(Service<HttpRequest, HttpResponse> delegate) {
-        super(delegate);
-    }
-
-    /**
-     * Determine if {@code request} is authorized for this service. If the result resolves to
-     * {@code true}, the request is authorized, or {@code false} otherwise. If the future
-     * resolves exceptionally, the request will not be authorized.
-     */
-    protected abstract CompletionStage<Boolean> authorize(HttpRequest request, ServiceRequestContext ctx);
-
-    /**
-     * Invoked when {@code req} is successful. By default, this method delegates the specified {@code req} to
-     * the {@link #delegate()} of this service.
-     */
-    protected HttpResponse onSuccess(ServiceRequestContext ctx, HttpRequest req) throws Exception {
-        return delegate().serve(ctx, req);
-    }
+    private final Authorizer<HttpRequest> authorizer;
+    private final AuthSuccessHandler<HttpRequest, HttpResponse> successHandler;
+    private final AuthFailureHandler<HttpRequest, HttpResponse> failureHandler;
 
-    /**
-     * Invoked when {@code req} is failed. By default, this method responds with the
-     * {@link HttpStatus#UNAUTHORIZED} status.
-     */
-    protected HttpResponse onFailure(ServiceRequestContext ctx, HttpRequest req, @Nullable Throwable cause)
-            throws Exception {
-        if (cause != null) {
-            logger.warn("Unexpected exception during authorization.", cause);
-        }
-        return HttpResponse.of(HttpStatus.UNAUTHORIZED);
+    HttpAuthService(Service<HttpRequest, HttpResponse> delegate, Authorizer<HttpRequest> authorizer,
+                    AuthSuccessHandler<HttpRequest, HttpResponse> successHandler,
+                    AuthFailureHandler<HttpRequest, HttpResponse> failureHandler) {
+        super(delegate);
+        this.authorizer = authorizer;
+        this.successHandler = successHandler;
+        this.failureHandler = failureHandler;
     }
 
     @Override
     public HttpResponse serve(ServiceRequestContext ctx, HttpRequest req) throws Exception {
-        return HttpResponse.from(authorize(req, ctx).handleAsync((result, t) -> {
+        return HttpResponse.from(authorize(ctx, req).handleAsync((result, t) -> {
             try {
                 if (t != null || !result) {
```

Comment:
Here we're not checking `result != null`

Reply:
Fixed